### DISCUSSION TITLE
react-query with axios post

### DISCUSSION THREAD

**Discussion by cseas** [TanStack/query#801] [General] [answered]
**Use case:**
I'm making an axios post call using redux-saga and storing my value in the redux store because the response is something I'd like to be available throughout the react app. I read some online discussions saying react-query can be used to replace redux so here I am experimenting but can't find any clues to using axios in the documentation.

I started with SWR and landed on their very minimal (not useful) [axios example](https://swr.vercel.app/docs/data-fetching#axios). Something like that would be a great addition to the react-query docs nonetheless.

Anyway, after a [long discussion](https://github.com/vercel/swr/issues/542#issuecomment-662164128), I finally found out that using SWR is actually not recommended for post calls. I was just wondering if that is the same case with react-query since both libraries seem to work pretty similarly.

**Comment by wolverineks** [comment, marked as answer]
https://react-query.tanstack.com/docs/guides/mutations#basic-mutations

**Reply by cseas** [reply]
Not sure how this answers the question. Can you elaborate a bit?

Is it fine to use react-query for POST calls or not? And how do I use axios with it?

**Reply by wolverineks** [reply]
`useQuery` is for `get`s. `useMutation` is for `post`s
It can handle anything that returns a promise.
Check out https://react-query.tanstack.com/docs/examples/auto-refetching

**Comment by banagale** [comment]
@wolverineks [I was looking to use](https://github.com/tannerlinsley/react-query/discussions/1107) useQuery() for a simple post as was done in [this example](https://github.com/tannerlinsley/react-query/blob/master/examples/custom-hooks/src/hooks/usePost.js). 

From what I can tell, useQuery is fine for simple POSTs. Can you give feedback on what kind of cutoff there is from when useQuery would not be advisable and useMutation should be used instead?

**Reply by cseas** [reply]
Hi @banagale, it looks like [the examples](https://github.com/tannerlinsley/react-query/tree/master/examples/custom-hooks/src/hooks) you linked to are using `GET` calls since you can see they're built with `axios.get()`.

A bit of confusion in words here perhaps. The function you're referring to in the examples is a function to fetch a "post", as in a website article "post", and not an HTTP POST. This is clear because the file right next to the one you linked to is `usePosts`.

I found react-query confusing and went back to Apollo because I had to pick something for my project. But it looks like the comments posted by @wolverineks are correct. I'll mark that comment as the answer.

**Reply by banagale** [reply]
Hey @cseas thanks for the kind reply. It is confusing having post in the get. 

In this case I was additionally confused by the argument being passed in, which I see now was being used to construct the url in that example. 

Fwiw, with some help I was able to get useQuery to perform a POST in my other discussion I linked above. 

I am still not sure on the why or best practices though.  

Thanks again for your reply.

**Reply by tannerlinsley** [reply]
Yeah, there's nothing stopping you from using any HTTP methods or even GRPC to "query" or "mutate" your data since React Query is based on just Promises. As far as naming things, this just comes down to preference. If you would rather see a different name for the hooks and models (blog posts) that the examples work with, feel free to open a PR :)

**Reply by banagale** [reply]
> Yeah, there's nothing stopping you from using any HTTP methods or even GRPC to "query" or "mutate" your data since React Query is based on just Promises. 

@tannerlinsley Thanks for this. Perhaps I'm overthinking this, I'm new to react and react query. 

However, I'm still not clear on when useQuery should be used and when useMutate should be used.

Would you please share a broad distinction between when useMutate should be used for requests that involve changes on the server?

 - [Docs on useQuery](https://react-query.tanstack.com/docs/guides/queries)
 - [Docs on useMutate](https://react-query.tanstack.com/docs/guides/mutations)

**Reply by wolverineks** [reply]
> > Yeah, there's nothing stopping you from using any HTTP methods or even GRPC to "query" or "mutate" your data since React Query is based on just Promises.
> 
> @tannerlinsley Thanks for this. Perhaps I'm overthinking this, I'm new to react and react query.
> 
> However, I'm still not clear on when useQuery should be used and when useMutate should be used.
> 
> Would you please share a broad distinction between when useMutate should be used for requests that involve changes on the server?
> 
> * [Docs on useQuery](https://react-query.tanstack.com/docs/guides/queries)
> * [Docs on useMutate](https://react-query.tanstack.com/docs/guides/mutations)

`useQuery` caches the data, making it available across multiple components
`useMutation` does not

**Reply by wolverineks** [reply]
`useQuery` could potentially not execute the request, and therefore not mutate the server state

**Reply by banagale** [reply]
> - `useQuery` caches the data, making it available across multiple components,`useMutation` does not
> - `useQuery` could potentially not execute the request, and therefore not mutate the server state

Thank you for these.

**Comment by tannerlinsley** [comment]
useQuery for reading data
useMutation for creating, changing, or  deleting data

Tanner Linsley

**Reply by banagale** [reply]
Thank you, I saw this indicated in the docs, but did not understand why. Contrasting these two with @wolverineks's details would be a helpful. Something like, 

**When to use useMutate and useQuery**
- useQuery is purely for reading data
- useMutation purely for creating, changing, or deleting data

**Why?**
- `useQuery` caches the data, making it available across multiple components,`useMutation` does not
- `useQuery` could potentially not execute the request, and therefore not mutate the server state
